crates/cli-support/src/testing.rs

```diff
@@ -0,0 +1,526 @@
+    }
+
+    pub fn with_format(mut self, format: OutputFormatSetting) -> Self {
+        self.format = Some(format);
+        self
+    }
+
+    pub fn with_filter<S: Into<String>>(mut self, filter: S) -> Self {
+        self.filter = Some(filter.into());
+        self
+    }
+
+    pub fn execute(&self) -> anyhow::Result<()> {
+        let shell = shell::Shell::new();
+
+        let file_name = self
+            .file
+            .file_name()
+            .map(Path::new)
+            .context("file to test is not a valid file, can't extract file name")?;
+
+        // Collect all tests that the test harness is supposed to run. We assume
+        // that any exported function with the prefix `__wbg_test` is a test we need
+        // to execute.
+        let wasm = fs::read(&self.file).context("failed to read Wasm file")?;
+        let mut wasm = walrus::ModuleConfig::new()
+            // generate dwarf by default, it can be controlled by debug profile
+            //
+            // https://doc.rust-lang.org/cargo/reference/profiles.html#debug
+            .generate_dwarf(true)
+            .parse(&wasm)
+            .context("failed to deserialize Wasm module")?;
+        let mut tests = Tests::new();
+
+        'outer: for export in wasm.exports.iter() {
+            let Some(name) = export.name.strip_prefix("__wbgt_") else {
+                continue;
+            };
+            let modifiers = name.split_once('_').expect("found invalid identifier").0;
+
+            let Some(name) = export.name.split_once("::").map(|s| s.1) else {
+                continue;
+            };
+
+            let test = Test {
+                name: name.into(),
+                export: export.name.clone(),
+                ignored: modifiers.contains('$'),
+            };
+
+            if let Some(filter) = &self.filter {
+                let matches = if self.exact {
+                    name == *filter
+                } else {
+                    name.contains(filter)
+                };
+
+                if !matches {
+                    tests.filtered += 1;
+                    continue;
+                }
+            }
+
+            for skip in &self.skip {
+                let matches = if self.exact {
+                    name == *skip
+                } else {
+                    name.contains(skip)
+                };
+
+                if matches {
+                    tests.filtered += 1;
+                    continue 'outer;
+                }
+            }
+
+            if !test.ignored && self.ignored {
+                tests.filtered += 1;
+            } else {
+                tests.tests.push(test);
+            }
+        }
+
+        if self.list {
+            for test in tests.tests {
+                println!("{}: test", test.name);
+            }
+
+            // Returning cleanly has the strange effect of outputting
+            // an additional empty line with spaces in it.
+            std::process::exit(0);
```

Comment:
We will have to do this in the runner instead.

Reply:
This is missing in the new runner binary now.

Reply:
If `execute` returns then the process would exit with `0`, right?

Reply:
Yes, but see the comment. We don't want those empty lines at the end because they break parsers like `cargo-nextest`.

Reply:
Yes.

Basically, by returning regularly form `fn main()`, e.g. via `Ok(())`, an extra empty line is inserted into the console. I have no idea why. But that extra line breaks tools trying to parse the output.

So we can just always return via `process::exit(0)` when `execute()` returns `Ok(())` to resolve the issue.

Reply:
Done